presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/ShardCompactionManager.java

```diff
@@ -180,47 +180,42 @@ private void discoverShards()
                 continue;
             }
 
-            Optional<Long> temporalColumnId = Optional.ofNullable(metadataDao.getTemporalColumnId(tableId));
-            if (temporalColumnId.isPresent() && !isValidTemporalColumn(tableId, temporalColumnId.get())) {
-                continue;
-            }
+            Table table = metadataDao.getTableInformation(tableId);
 
-            CompactionSetCreator compactionSetCreator = getCompactionSetCreator(tableId, temporalColumnId);
-            Set<ShardMetadata> shards = getFilteredShards(entry.getValue(), tableId, temporalColumnId);
+            CompactionSetCreator compactionSetCreator = getCompactionSetCreator(table);
+            Set<ShardMetadata> shards = getFilteredShards(entry.getValue(), table);
 
             addToCompactionQueue(compactionSetCreator, tableId, shards);
         }
     }
 
-    private Set<ShardMetadata> getFilteredShards(List<ShardMetadata> shardMetadatas, long tableId, Optional<Long> temporalColumnId)
+    private Set<ShardMetadata> getFilteredShards(List<ShardMetadata> shardMetadatas, Table tableInfo)
     {
         Set<ShardMetadata> shards = shardMetadatas.stream()
                 .filter(this::needsCompaction)
                 .filter(shard -> !organizer.inProgress(shard.getShardUuid()))
                 .collect(toSet());
 
-        if (temporalColumnId.isPresent()) {
-            shards = filterShardsWithTemporalMetadata(shards, tableId, temporalColumnId.get());
+        if (tableInfo.getTemporalColumnId().isPresent()) {
+            shards = filterShardsWithTemporalMetadata(shards, tableInfo.getTableId(), tableInfo.getTemporalColumnId().getAsLong());
         }
 
         return shards;
     }
 
-    private CompactionSetCreator getCompactionSetCreator(long tableId, Optional<Long> temporalColumnId)
+    private CompactionSetCreator getCompactionSetCreator(Table tableInfo)
     {
-        if (!temporalColumnId.isPresent()) {
+        if (!tableInfo.getTemporalColumnId().isPresent()) {
             return new FileCompactionSetCreator(maxShardSize, maxShardRows);
         }
 
-        checkState(isValidTemporalColumn(tableId, temporalColumnId.get()), "invalid temporal column type");
-        Type type = metadataDao.getTableColumn(tableId, temporalColumnId.get()).getDataType();
-
+        Type type = metadataDao.getTableColumn(tableInfo.getTableId(), tableInfo.getTemporalColumnId().getAsLong()).getDataType();
```

Comment:
We could do a `LEFT JOIN` in the query and get the type at the same time. (just an idea to consider)